### PR TITLE
Added test cases for Enumerable methods

### PR DESCRIPTION
**Pull request by Maclenn77**
Hi! @RossielCS and @Maclenn77 here.

We added test cases for the enumerable methods. This worked to find some issues that we didn't see before.

### Tested Methods

- my_each
- my_each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

In total, they are 34 test cases.

**Thanks for reviewing!**

### CODE REVIEW

File: spec/main_spec.rb [hunk 1]
```diff
@@ -1,2 +1,167 @@
...
+
+  describe '#my_each' do
+    it 'returns an enumerator if the block is not given' do
+      expect(array.my_each).to be_an_instance_of(Enumerator)
+    end
+
+    it 'returns the element if the block is given' do
+      expect(array.my_each { |x| x * 2 }).to eql(array)
+    end
+  end
```
Comment (shshamim63) [inline spec/main_spec.rb]: [Optional]
You can also add a test to check, it is calling the block for each of the items. And as it is derived from an actual method it will be better to compare it with the actual each. Now the question is how can we do that? Check out the sample test below.
```ruby
it 'calls the given block for each of the items in self' do
      each_output = 0
      block = proc { |num| each_output += num }
      array.my_each(&block)
      my_each_output = each_output.dup
      each_output = 0
      array.each(&block)
      expect(my_each_output).to eq(each_output)
end
```

File: spec/main_spec.rb [hunk 2]
```diff
@@ -1,2 +1,167 @@
...
+
+  describe '#my_each_with_index' do
+    it 'returns an enumerator if the block is not given' do
+      expect(array.my_each_with_index).to be_an_instance_of(Enumerator)
+    end
+
+    it 'returns the element if the block is given' do
+      expect(array.my_each_with_index { |x, _| x * 2 }).to eql(array)
+    end
+  end
```
Comment (shshamim63) [inline spec/main_spec.rb]: [Optional]
You can also add a test similar to the given sample test above.

File: spec/main_spec.rb [hunk 3]
```diff
@@ -1,2 +1,167 @@
...
+    end
+
+    it 'when block is given, it doesn\'t return true if at least one case is false' do
+      expect(array.my_all? { |m| m > 1 }).not_to eql(true)
+    end
+
+    it 'when block is given, it returns true if all cases are true in a range' do
+      expect(range.my_all? { |m| m.is_a? Integer }).to eql(true)
+    end
+  end
```
Comment (shshamim63) [inline spec/main_spec.rb]: I cannot see test for the following edge cases of my_all? Please add them
- When a regex is passed as an argument.
- when a value is given other than a class or regex as an argument

File: spec/main_spec.rb [hunk 4]
```diff
@@ -1,2 +1,167 @@
...
+    end
+
+    it 'returns true if pattern === element is true for an element' do
+      expect(array_boolean.my_any?(nil)).to eql(true)
+    end
+
+    it 'doesn\'t returns true if pattern === element is false for all elements' do
+      expect(array_false.my_any?(Integer)).not_to eql(true)
+    end
+  end
```
Comment (shshamim63) [inline spec/main_spec.rb]: - Add a test case related to regex.

File: spec/main_spec.rb [hunk 5]
```diff
@@ -1,2 +1,167 @@
...
+    end
+
+    it 'returns true if pattern === element is false for all elements' do
+      expect(array.my_none?(Float)).to eql(true)
+    end
+
+    it 'doesn\'t returns true if pattern === element is true for an element' do
+      expect(array_boolean.my_none?(nil)).not_to eql(true)
+    end
+  end
```
Comment (shshamim63) [inline spec/main_spec.rb]: - Add a test case related to regex as an argument.

File: spec/main_spec.rb [hunk 6]
```diff
@@ -1,2 +1,167 @@
...
+    it 'returns total items of an array if block is not given' do
+      expect(array.my_count).to eql(6)
+    end
+
+    it 'returns of characters of a string given like argument' do
+      class String
+        include Enumerable
+      end
+      expect(string.my_count('n')).to eql(1)
+    end
```
Comment (shshamim63) [inline spec/main_spec.rb:182]: Wow, clever one. :tada:

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @Maclenn77 and @RossielCS :wave: 
Awesome job :clap: Almost covered all the test cases, but I can see some space where you need to work a bit. Follow the comment
___
### :spades: Mandatory Changes :spades: 
___
1. Add some tests to the following methods.
- my_all?
- my_any?
- my_none?
2. I cannot see any info into the readme related to test at least how to execute the test. Maybe you can add
To execute the test perform the following command in your terminal.
```bash
rspec spec/main_spec.rb -fdoc #it will provide you well-organized output on your terminal
```
___
### :clubs: Optional Changes :clubs: 
___
You can test your methods by comparing the actual one. Check the sample test to get an idea.
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Review by AdnanAfsari — APPROVED** [review]
**Status: Approved**

Hello @Maclenn77 @RossielCS ✋ 

Nice work, I think you have made all the changes requested by the previous reviewer 👏 
It's a 👍 from me.

Happy Coding
Adnan Afsari
